# Task 4.8: Grouping Data & Aggregating Variables

## Question 1: Importations

In [1]:
# Import Libraries 
import pandas as pd 
import numpy as np 
import os

In [2]:
# Set Path 
path = r'C:\Users\calli\OneDrive\Data Immersion\Achievement 4' 

In [17]:
# Import orders_products_merged_3
ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged_2.pkl')) 

In [18]:
# Show database
ords_prods_merge.head()

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2398795        1             2                  3                  7   
2    473747        1             3                  3                 12   
3   2254736        1             4                  4                  7   
4    431534        1             5                  4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered _merge  \
0                     NaN         196                  1          0   both   
1                    15.0         196                  1          1   both   
2                    21.0         196                  1          1   both   
3                    29.0         196                  1          1   both   
4                    28.0         196                  1          1   both   

  product_name  aisle_id  department_id  prices        price_label  \
0         Soda        77              7     9.0  Mid-range product   
1         Soda        77              7     9.0  Mid-range product   
2         Soda        77              7     9.0  Mid-range product   
3         Soda        77              7     9.0  Mid-range product   
4         Soda        77              7     9.0  Mid-range product   

      busiest_day    busiest_days busiest_period_of_day  
0  Regularly busy  Regularly busy        Average Orders  
1  Regularly busy      Least busy        Average Orders  
2  Regularly busy      Least busy        Average Orders  
3      Least busy      Least busy        Average Orders  
4      Least busy      Least busy        Average Orders

## Question 2: 

In [19]:
# Find the aggregated mean of the “order_number” column grouped by “department_id”
ords_prods_merge.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
1                15.457838
2                17.277920
3                17.170395
4                17.811403
5                15.215751
6                16.439806
7                17.225802
8                15.340650
9                15.895474
10               20.197148
11               16.170638
12               15.887671
13               16.583536
14               16.773669
15               16.165037
16               17.665606
17               15.694469
18               19.310397
19               17.177343
20               16.473447
21               22.902379

## Question 3:

### 1) Department 21 has the highest average order number. (22.9)

### 2) Department 5 has the lowest average order number. (15.2)

### 3) There are 21 total departments, which is more than there were in the subset from the exercise.

## Question 4: Create a loyalty flag

### Use the Transform Function

In [20]:
# Use the transform function to create a new column
ords_prods_merge['max_order'] = ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)

In [22]:
# Show results of the transform function
ords_prods_merge.head(60)

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0    2539329        1             1                  2                  8   
1    2398795        1             2                  3                  7   
2     473747        1             3                  3                 12   
3    2254736        1             4                  4                  7   
4     431534        1             5                  4                 15   
5    3367565        1             6                  2                  7   
6     550135        1             7                  1                  9   
7    3108588        1             8                  1                 14   
8    2295261        1             9                  1                 16   
9    2550362        1            10                  4                  8   
10   2968173       15            15                  1                  9   
11   1870022       15            17                  2                 16   
12   1911383       15            18                  2                 11   
13   2715276       15            21                  1                  9   
14    487368       15            22                  1                 10   
15   2293453       19             2                  5                 14   
16   1973799       19             5                  6                 12   
17    532817       19             7                  4                 17   
18   1573906       21            10                  3                 10   
19   1593000       31            10                  3                  8   
20   2231262       31            17                  3                 11   
21   2580647       43             6                  4                 16   
22   2187180       43             9                  4                 12   
23   2497897       52             1                  1                  9   
24   1318871       52             2                  1                 10   
25   1261384       52             3                  1                 10   
26    580568       52             5                  2                 10   
27   2428073       52             7                  2                  7   
28    342306       52             9                  1                 11   
29   1520353       52            10                  2                 10   
30   2542086       52            11                  2                  9   
31    944694       52            12                  1                 16   
32   1268191       52            15                  2                 11   
33   1122089       52            17                  2                 10   
34   1498922       52            18                  2                 10   
35    180919       52            19                  3                  8   
36   2695875       52            21                  5                 10   
37   3205960       67             1                  3                 11   
38    970922       67             2                  1                 15   
39    596585       67             3                  1                 13   
40   2037087       67             4                  2                  9   
41   2474111       67             5                  3                 14   
42   1568481       67             6                  3                 12   
43   2048527       67             8                  2                 11   
44   2389031       67             9                  2                 14   
45   2378562       67            10                  1                 10   
46   1515698       67            11                  1                 11   
47   1928050       67            12                  2                 12   
48   1949274       67            13                  1                 10   
49    551869       67            14                  1                  9   
50    768572       67            16                  4                 10   


## Use the Loc Function

In [23]:
# Flag customers with over 40 orders as 'Loyal customer' 
ords_prods_merge.loc[ords_prods_merge['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer' 

In [24]:
# Flag customers with more than 10 and up to 40 orders as 'Regular customer' 
ords_prods_merge.loc[(ords_prods_merge['max_order'] <= 40) & (ords_prods_merge['max_order'] > 10), 'loyalty_flag'] = 'Regular customer' 

In [25]:
# Flag customers with 10 or fewer orders as 'New customer' 
ords_prods_merge.loc[ords_prods_merge['max_order'] <= 10, 'loyalty_flag'] = 'New customer' 

In [26]:
# Check value counts of loyalty flag 
ords_prods_merge['loyalty_flag'].value_counts(dropna = False) 

loyalty_flag
Regular customer    15876776
Loyal customer      10284093
New customer         6243990
Name: count, dtype: int64

In [27]:
# Check head of certain columns
ords_prods_merge[['user_id', 'loyalty_flag', 'order_number']].head(60)

user_id      loyalty_flag  order_number
0         1      New customer             1
1         1      New customer             2
2         1      New customer             3
3         1      New customer             4
4         1      New customer             5
5         1      New customer             6
6         1      New customer             7
7         1      New customer             8
8         1      New customer             9
9         1      New customer            10
10       15  Regular customer            15
11       15  Regular customer            17
12       15  Regular customer            18
13       15  Regular customer            21
14       15  Regular customer            22
15       19      New customer             2
16       19      New customer             5
17       19      New customer             7
18       21  Regular customer            10
19       31  Regular customer            10
20       31  Regular customer            17
21       43  Regular customer             6
22       43  Regular customer             9
23       52  Regular customer             1
24       52  Regular customer             2
25       52  Regular customer             3
26       52  Regular customer             5
27       52  Regular customer             7
28       52  Regular customer             9
29       52  Regular customer            10
30       52  Regular customer            11
31       52  Regular customer            12
32       52  Regular customer            15
33       52  Regular customer            17
34       52  Regular customer            18
35       52  Regular customer            19
36       52  Regular customer            21
37       67  Regular customer             1
38       67  Regular customer             2
39       67  Regular customer             3
40       67  Regular customer             4
41       67  Regular customer             5
42       67  Regular customer             6
43       67  Regular customer             8
44       67  Regular customer             9
45       67  Regular customer            10
46       67  Regular customer            11
47       67  Regular customer            12
48       67  Regular customer            13
49       67  Regular customer            14
50       67  Regular customer            16
51       67  Regular customer            18
52       67  Regular customer            19
53       67  Regular customer            21
54       67  Regular customer            22
55       67  Regular customer            23
56       81      New customer             3
57       81      New customer             5
58       82  Regular customer             1
59       82  Regular customer             3

## Question 5: Check statistics using the loyalty flag

In [30]:
# Show head
ords_prods_merge.head()

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2398795        1             2                  3                  7   
2    473747        1             3                  3                 12   
3   2254736        1             4                  4                  7   
4    431534        1             5                  4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered _merge  \
0                     NaN         196                  1          0   both   
1                    15.0         196                  1          1   both   
2                    21.0         196                  1          1   both   
3                    29.0         196                  1          1   both   
4                    28.0         196                  1          1   both   

  product_name  aisle_id  department_id  prices        price_label  \
0         Soda        77              7     9.0  Mid-range product   
1         Soda        77              7     9.0  Mid-range product   
2         Soda        77              7     9.0  Mid-range product   
3         Soda        77              7     9.0  Mid-range product   
4         Soda        77              7     9.0  Mid-range product   

      busiest_day    busiest_days busiest_period_of_day  max_order  \
0  Regularly busy  Regularly busy        Average Orders         10   
1  Regularly busy      Least busy        Average Orders         10   
2  Regularly busy      Least busy        Average Orders         10   
3      Least busy      Least busy        Average Orders         10   
4      Least busy      Least busy        Average Orders         10   

   loyalty_flag  
0  New customer  
1  New customer  
2  New customer  
3  New customer  
4  New customer

In [45]:
# Group by loyalty_flag showing statistics of prices
ords_prods_merge.groupby('loyalty_flag').agg({'prices': ['count', 'mean', 'min', 'max', 'std', 'median']})

prices                                            
                     count       mean  min      max         std median
loyalty_flag                                                          
Loyal customer    10284093  10.386336  1.0  99999.0  328.017787    7.4
New customer       6243990  13.294670  1.0  99999.0  597.560299    7.4
Regular customer  15876776  12.495717  1.0  99999.0  539.720919    7.4

### Question 5 Observations:

#### 1) The standard deviation of loyal customers is much lower than new and regular customers. New customers have the highest.

#### 2) The loyal customers have the lowest average, followed by regular, and new customers have the highest.

#### 3) The minimum, median, and maximum values are the same across all flags.

## Question 6: Create a spending flag

In [35]:
# Use the transform function to create a new column of mean prices
ords_prods_merge['mean_order'] = ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)

In [40]:
# Show head
ords_prods_merge.head()

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2398795        1             2                  3                  7   
2    473747        1             3                  3                 12   
3   2254736        1             4                  4                  7   
4    431534        1             5                  4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered _merge  \
0                     NaN         196                  1          0   both   
1                    15.0         196                  1          1   both   
2                    21.0         196                  1          1   both   
3                    29.0         196                  1          1   both   
4                    28.0         196                  1          1   both   

   ... aisle_id  department_id  prices        price_label     busiest_day  \
0  ...       77              7     9.0  Mid-range product  Regularly busy   
1  ...       77              7     9.0  Mid-range product  Regularly busy   
2  ...       77              7     9.0  Mid-range product  Regularly busy   
3  ...       77              7     9.0  Mid-range product      Least busy   
4  ...       77              7     9.0  Mid-range product      Least busy   

     busiest_days busiest_period_of_day max_order  loyalty_flag average_prices  
0  Regularly busy        Average Orders        10  New customer       6.367797  
1      Least busy        Average Orders        10  New customer       6.367797  
2      Least busy        Average Orders        10  New customer       6.367797  
3      Least busy        Average Orders        10  New customer       6.367797  
4      Least busy        Average Orders        10  New customer       6.367797  

[5 rows x 21 columns]

In [39]:
# change mean_order column name to average_prices
ords_prods_merge.rename(columns={'mean_order': 'average_prices'}, inplace = True)

In [41]:
# Flag customers with a mean product price of under 10 as 'Low spender' 
ords_prods_merge.loc[ords_prods_merge['average_prices'] < 10, 'spending_flag'] = 'Low spender'

In [42]:
# Flag customers with a mean product price of 10 or more as 'High spender' 
ords_prods_merge.loc[ords_prods_merge['average_prices'] >= 10, 'spending_flag'] = 'High spender'

In [43]:
# Show value counts of high and low spenders
ords_prods_merge['spending_flag'].value_counts(dropna = False)

spending_flag
Low spender     31770614
High spender      634245
Name: count, dtype: int64

In [44]:
# Show head
ords_prods_merge.head()

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2398795        1             2                  3                  7   
2    473747        1             3                  3                 12   
3   2254736        1             4                  4                  7   
4    431534        1             5                  4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered _merge  \
0                     NaN         196                  1          0   both   
1                    15.0         196                  1          1   both   
2                    21.0         196                  1          1   both   
3                    29.0         196                  1          1   both   
4                    28.0         196                  1          1   both   

   ... department_id  prices        price_label     busiest_day  \
0  ...             7     9.0  Mid-range product  Regularly busy   
1  ...             7     9.0  Mid-range product  Regularly busy   
2  ...             7     9.0  Mid-range product  Regularly busy   
3  ...             7     9.0  Mid-range product      Least busy   
4  ...             7     9.0  Mid-range product      Least busy   

     busiest_days busiest_period_of_day max_order  loyalty_flag  \
0  Regularly busy        Average Orders        10  New customer   
1      Least busy        Average Orders        10  New customer   
2      Least busy        Average Orders        10  New customer   
3      Least busy        Average Orders        10  New customer   
4      Least busy        Average Orders        10  New customer   

   average_prices spending_flag  
0        6.367797   Low spender  
1        6.367797   Low spender  
2        6.367797   Low spender  
3        6.367797   Low spender  
4        6.367797   Low spender  

[5 rows x 22 columns]

## Question 7: Create an order frequency flag

In [46]:
# Use the transform function to create a new column of median days since prior order
ords_prods_merge['median_days_since_prior_order'] = ords_prods_merge.groupby(['user_id'])['days_since_prior_order'].transform(np.median)

In [48]:
# show head
ords_prods_merge.head()

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2398795        1             2                  3                  7   
2    473747        1             3                  3                 12   
3   2254736        1             4                  4                  7   
4    431534        1             5                  4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered _merge  \
0                     NaN         196                  1          0   both   
1                    15.0         196                  1          1   both   
2                    21.0         196                  1          1   both   
3                    29.0         196                  1          1   both   
4                    28.0         196                  1          1   both   

   ... prices        price_label     busiest_day    busiest_days  \
0  ...    9.0  Mid-range product  Regularly busy  Regularly busy   
1  ...    9.0  Mid-range product  Regularly busy      Least busy   
2  ...    9.0  Mid-range product  Regularly busy      Least busy   
3  ...    9.0  Mid-range product      Least busy      Least busy   
4  ...    9.0  Mid-range product      Least busy      Least busy   

  busiest_period_of_day max_order  loyalty_flag average_prices  spending_flag  \
0        Average Orders        10  New customer       6.367797    Low spender   
1        Average Orders        10  New customer       6.367797    Low spender   
2        Average Orders        10  New customer       6.367797    Low spender   
3        Average Orders        10  New customer       6.367797    Low spender   
4        Average Orders        10  New customer       6.367797    Low spender   

  median_days_since_prior_order  
0                          20.5  
1                          20.5  
2                          20.5  
3                          20.5  
4                          20.5  

[5 rows x 23 columns]

In [49]:
# Flag customers with a median of over 20 days since their prior order as 'Non-frequent customer'
ords_prods_merge.loc[ords_prods_merge['median_days_since_prior_order'] > 20, 'frequency_flag'] = 'Non-frequent customer'

In [50]:
# Flag customers with a median of over 10 and 20 or fewer days since their prior order as 'Regular customer'
ords_prods_merge.loc[(ords_prods_merge['median_days_since_prior_order'] > 10) & (ords_prods_merge['median_days_since_prior_order'] <= 20), 'frequency_flag'] = 'Regular customer'

In [51]:
# Flag customers with a median of 10 or fewer days since their prior order as 'Frequent customer'
ords_prods_merge.loc[ords_prods_merge['median_days_since_prior_order'] <= 10, 'frequency_flag'] = 'Frequent customer'

In [52]:
# Check value counts
ords_prods_merge['frequency_flag'].value_counts(dropna = False)

frequency_flag
Frequent customer        21559853
Regular customer          7208564
Non-frequent customer     3636437
NaN                             5
Name: count, dtype: int64

## Question 9: Export dataframe as pickle file

In [53]:
# Export data to pkl 

ords_prods_merge.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_merged_4.pkl')) 